src/bn254.ts

```diff
@@ -1,234 +1,209 @@
-    return [["uint256"], [this.f.serialize()]];
+  abiEncode(): ABIEncoded {
+    return [["uint256"], [this.toEvm()]];
   }
 
   add(e: Fr): this {
-    this.f = mcl.add(this.f, e.f);
+    this.f = IFr.add(this.f, e.f);
     return this;
   }
 
   mul(e: Fr): this {
-    this.f = mcl.mul(this.f, e.f);
+    this.f = IFr.mul(this.f, e.f);
     return this;
   }
 
   neg(): this {
-    this.f = mcl.neg(this.f);
+    this.f = IFr.neg(this.f);
     return this;
   }
 
   inverse(): this {
-    this.f = mcl.inv(this.f);
+    this.f = IFr.inv(this.f);
     return this;
   }
 
   zero(): this {
-    this.f.setInt(0);
+    this.f = IFr.zero;
     return this;
   }
 
   random(): this {
-    this.f.setByCSPRNG();
+    this.f = IFr.random();
     return this;
   }
 
   one(): this {
-    this.f.setInt(1);
+    this.f = IFr.one;
     return this;
   }
 
   equal(e: Fr): boolean {
-    return this.f.isEqual(e.f);
+    return IFr.eq(this.f, e.f);
   }
 
   serialize(): Uint8Array {
-    return this.f.serialize();
+    const buff = new Uint8Array(32);
+    IFr.toRprLE(buff, 0, this.f);
+    return buff;
   }
 
   deserialize(buff: Uint8Array): EncodingRes<this> {
-    this.f.deserialize(buff);
-    // ??? no valid functions ???
+    this.f = IFr.fromRprLE(buff, 0);
     return ok(this);
   }
 
-  fromBytes(array: Uint8Array): this {
-    this.f.setLittleEndianMod(array);
-    return this;
-  }
-
   set(e: Fr): this {
-    const arr = e.f.serialize();
+    const arr = e.serialize();
     this.deserialize(arr)._unsafeUnwrap();
     return this;
   }
 
   toEvm(): BigNumber {
-    return arrayToBn(this.f.serialize());
+    const obj = IFr.toObject(this.f);
+    return BigNumber.from(obj);
   }
 
   fromEvm(t: BigNumber): EncodingRes<this> {
-    this.f.deserialize(bnToArray(t));
+    this.f = IFr.fromObject(t.toBigInt());
     return ok(this);
   }
 }
 
 export class G1
   implements Point<Fr>, Atom<Fr>, EVMEncoding<EVMG1Point>, ABIEncoder, ToBytes
 {
-  p: mcl.G1;
+  p: Uint8Array;
+
   constructor() {
-    this.p = new mcl.G1();
+    this.p = IG1.g;
   }
 
-  abiEncode(): [Array<string>, Array<any>] {
+  abiEncode(): ABIEncoded {
     const evm = this.toEvm();
-    const xarray = bnToArray(evm.x, false, 32);
-    const yarray = bnToArray(evm.y, false, 32);
     return [
       ["uint256", "uint256"],
-      [xarray, yarray],
+      [evm.x, evm.y],
     ];
   }
 
   add(e: G1): this {
-    this.p = mcl.add(this.p, e.p);
+    this.p = IG1.add(this.p, e.p);
     return this;
   }
 
   mul(e: Fr): this {
-    this.p = mcl.mul(this.p, e.f);
+    this.p = IG1.timesFr(this.p, e.f);
     return this;
   }
 
   neg(): this {
-    this.p = mcl.neg(this.p);
+    this.p = IG1.neg(this.p);
     return this;
   }
 
   zero(): this {
+    this.p = IG1.zero;
     return this;
   }
 
   random(): this {
-    const random = randHex(32);
-    return this.setHashOf(random);
-  }
-
-  setHashOf(m: string): this {
-    this.p.setHashOf(m);
+    const f = IFr.random();
```

Comment:
Please put a comment on the method or in here that this method is currently insecure (hashing should not be done like that - maybe it was like this before but it's something we'll need to fix).

Reply:
Yeah this was how it was implemented ... probably because the ffjavascript library doesn't support `setHashOf`. The diff is weird because the file was renamed from bn254_iden to bn254 (and the previous bn254 with mcl-wasm deleted)

Reply:
I think the diff looks weird ... I see this code, nothing related to hashing
```
random(): this {
    const f = IFr.random();
    this.p = IG1.timesFr(IG1.g, f);
    // this.p = IG1.fromRng(utils.getThreadRng());
    return this
}
```